File: apps/worker_whatsmeow/internal/app/message_status_identity.go
package app

import (
	"crypto/sha256"
	"encoding/hex"
	"strings"
)

const messageStatusEventIDVersion = "v1"

func messageStatusRevision(update *MessageStatusUpdate) string {
	if update == nil {
		return "unknown"
	}
	if update.Failed && update.Ambiguous {
		return "ambiguous"
	}
	if update.Failed {
		return "failed"
	}
	if value, _ := update.Patch["is_seen"].(bool); value {
		return "seen"
	}
	if value, _ := update.Patch["is_delivered"].(bool); value {
		return "delivered"
	}
	if value, _ := update.Patch["is_sent"].(bool); value {
		return "sent"
	}
	return "unknown"
}

func ensureMessageStatusEventID(update *MessageStatusUpdate) string {
	if update == nil {
		return ""
	}
	if existing := strings.TrimSpace(update.EventID); existing != "" {
		update.EventID = existing
		return existing
	}
	accountID := strings.TrimSpace(update.AccountID)
	workerID := strings.TrimSpace(update.WorkerID)
	messageID := canonicalInboundStanzaID(update.MessageID)
	if accountID == "" || workerID == "" || messageID == "" {
		return ""
	}
	canonical := strings.Join([]string{
		messageStatusEventIDVersion,
		accountID,
		workerID,
		messageID,
		messageStatusRevision(update),
	}, "\x00")
	digest := sha256.Sum256([]byte(canonical))
	update.EventID = "message_status_" + messageStatusEventIDVersion + "_" + hex.EncodeToString(digest[:])
	return update.EventID
}

func messageStatusKafkaKey(accountID, workerID, messageID string) string {
	return strings.Join([]string{
		strings.TrimSpace(accountID),
		strings.TrimSpace(workerID),
		canonicalInboundStanzaID(messageID),
	}, ":")
}

File: packages/common/functions/messageStatusIdentity.ts
import { createHash } from 'node:crypto';
import type { IMessageStatusUpdate } from '../interfaces/IMessageStatusUpdate';
import { parseSerializedMessageId } from './parseSerializedMessageId';

const MESSAGE_STATUS_EVENT_ID_VERSION = 'v1';

function nonEmpty(value?: string | null): string | null {
  const normalized = value?.trim();
  return normalized ? normalized : null;
}

export function canonicalMessageStatusMessageId(
  value?: string | null
): string | null {
  const messageId = nonEmpty(value);
  if (!messageId) {
    return null;
  }
  return parseSerializedMessageId(messageId)?.stanzaId ?? messageId;
}

function statusRevision(
  data: Pick<IMessageStatusUpdate, 'patch' | 'failed' | 'ambiguous'>
): string {
  if (data.failed === true && data.ambiguous === true) return 'ambiguous';
  if (data.failed === true) return 'failed';
  if (data.patch.is_seen === true) return 'seen';
  if (data.patch.is_delivered === true) return 'delivered';
  if (data.patch.is_sent === true) return 'sent';
  return 'unknown';
}

export function buildMessageStatusEventId(
  data: Pick<
    IMessageStatusUpdate,
    'account_id' | 'worker_id' | 'message_id' | 'patch' | 'failed' | 'ambiguous'
  >
): string | null {
  const accountId = nonEmpty(data.account_id);
  const workerId = nonEmpty(data.worker_id);
  const messageId = canonicalMessageStatusMessageId(data.message_id);
  if (!accountId || !workerId || !messageId) {
    return null;
  }

  const canonical = [
    MESSAGE_STATUS_EVENT_ID_VERSION,
    accountId,
    workerId,
    messageId,
    statusRevision(data),
  ].join('\0');

  return `message_status_${MESSAGE_STATUS_EVENT_ID_VERSION}_${createHash(
    'sha256'
  )
    .update(canonical)
    .digest('hex')}`;
}

export function ensureMessageStatusEventId(
  data: IMessageStatusUpdate
): string | null {
  const existing = nonEmpty(data.event_id);
  if (existing) {
    data.event_id = existing;
    return existing;
  }

  const generated = buildMessageStatusEventId(data);
  if (generated) {
    data.event_id = generated;
  }
  return generated;
}
